Pairwise postprocessing in validation for both: images and representations

AlekseySh opened the pull request:
* In addition to pairs of embeddings, added validation for pairs of images
* Refactored the old code

dapladoc commented on oml/retrieval/postprocessors/pairwise.py:
```diff
@@ -0,0 +1,206 @@
+
+        # 1. Adjust top_n with respect to the actual gallery size and find top-n pairs
+        top_n = min(self.top_n, n_galleries)
+        ii_top = torch.topk(distances, k=top_n, largest=False)[1].view(-1)
+
+        # 2. Create (n_queries * top_n) pairs of each query and related galleries and re-estimate distances for them
+        distances_upd = self.inference(queries=queries, galleries=galleries, ii_top=ii_top, top_n=top_n)
```
Here `ii_top` is a flattened array, but in the `self.inference` docstring it is said that `ii_top: Indices of the closest galleries with the shape of [Q, top_n]`. Is it ok to use flattened array in such case?

AlekseySh replied:
thx for your findings 
code is okay, but I will fix comments

AlekseySh replied:
thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`

dapladoc commented on oml/retrieval/postprocessors/pairwise.py:
```diff
@@ -0,0 +1,206 @@
+
+        """
+        queries = queries.repeat_interleave(top_n, dim=0)
+        galleries = galleries[ii_top]
```
If `ii_top` is of size `[Q, top_n]` then `galleries` will be of size `[Q, top_n, dim]`, where `dim` is the embeddings dimension. It doesn't look like it is supposed to be.

AlekseySh replied:
thx for your findings 
code is okay, but I will fix comments

AlekseySh replied:
thx for your findings
I've aligned shapes and comments for `ii_top` and `distances`

dapladoc commented on oml/inference/base.py:
```diff
@@ -0,0 +1,45 @@
+    if f_imread is None:
+        f_imread = get_im_reader_for_transforms(transform)
+
+    prev_mode = model.training
+    model.eval()
+
+    dataset = ListDataset(paths, bboxes=None, transform=transform, f_imread=f_imread)
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+    device = next(model.parameters()).device
```
The same thoughts here about introducing the `get_device` function, and the context manager to restore model training flag.

AlekseySh replied:
done

AlekseySh commented:
* added pairwise models (and corresponding pair datasets + inference code for them)
* added pairwise postprocessors (on embeddings and on images). organised code, so it's shared among both postprocessors 
* Added new models and postprocessors to the registry and docs
* added creating a postprocessor to the validation entry point (not to a training one yet)
* Added tests for postprocessor on images pairs

AlekseySh commented on oml/inference/pairwise.py:
```diff
@@ -0,0 +1,75 @@
+    prev_mode = model.training
+    model.eval()
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+    device = next(model.parameters()).device
```
done

AlekseySh commented on oml/inference/pairwise.py:
```diff
@@ -0,0 +1,80 @@
+        batch_size: int,
+        verbose: bool
+) -> Tensor:
+    prev_mode = model.training
```
The idea is good, but this PR is too big already (~800 lines). So, if you or someone else do it as a small separate PR I will be happy

AlekseySh replied:
okay, I added it in this PR

AlekseySh replied:
thx for a draft

dapladoc commented on oml/metrics/embeddings.py:
```diff
@@ -170,17 +172,16 @@ def _calc_matrices(self) -> None:
         if self.postprocessor is None:
             pass
         elif isinstance(self.postprocessor, PairwiseEmbeddingsPostprocessor):
-            max_k = max([*self.cmc_top_k, *self.precision_top_k, *self.map_top_k])
-            if max_k > self.postprocessor.top_n:
-                warn(
-                    f"One of retrieval metrics will be computed at k = {max_k},"
-                    f"but postprocessor will re-rank only {self.postprocessor.top_n} closest galleries."
-                    f"Make sure that this is the desired behaviour."
-                )
```
Sorry, I can't find where you decided to put this warning. Or did you get rid of them at all?

DaloroAT commented on docs/source/contents/postprocessing.rst:
```diff
@@ -1,48 +1,108 @@
+    .. automethod:: __init__
+    .. automethod:: __getitem__
+
+Implementations
```
readthedocs looks ugly, it should be nested

AlekseySh replied:
reworked

DaloroAT commented on oml/inference/base.py:
```diff
@@ -0,0 +1,43 @@
+    transform: TTransforms,
+    num_workers: int,
+    batch_size: int,
+    verbose: bool,
```
Suppose `verbose: bool = False` is ok

AlekseySh replied:
agree

DaloroAT commented on oml/inference/base.py:
```diff
@@ -0,0 +1,43 @@
+    model: nn.Module,
+    paths: List[Path],
+    transform: TTransforms,
+    num_workers: int,
```
`num_workers: int = 0` is also ok

AlekseySh replied:
I prefer to specify it directly because it's easy to forget
And then you may need a lot of time to notice that inference takes too long

AlekseySh commented on oml/metrics/embeddings.py:
```diff
@@ -170,17 +172,16 @@ def _calc_matrices(self) -> None:
+        elif isinstance(self.postprocessor, PairwiseImagesPostprocessor):
+            self.distance_matrix = self.postprocessor.process(
+                distances=self.distance_matrix,
+                queries=np.array(self.acc.storage[PATHS_KEY])[is_query],  # type: ignore
```
but we don't know in advance if our postprocessor need this key or not
I suggest to do it in the next pr when we introduce method `postprocessor.get_needed_keys()`

DaloroAT replied:
I mean that you need to introduce `needed_keys`, because right now it's hardcoded and you will fail only all computations

Ok, next PR

AlekseySh commented on oml/registry/pairwise_models.py:
```diff
@@ -0,0 +1,22 @@
+PAIRWISE_MODELS_REGISTRY = {"linear_siamese": LinearSiamese, "resnet_siamese": ResNetSiamese}
+
+
+def get_pairwise_model(model_name: str, **kwargs: Dict[str, Any]) -> IPairwiseModel:
```
fixed

AlekseySh commented on oml/utils/misc_torch.py:
```diff
@@ -95,6 +96,16 @@ def pairwise_dist(x1: Tensor, x2: Tensor, p: int = 2) -> Tensor:
 
 
+def normalise(x: Tensor, p: int = 2) -> Tensor:
+    xn = torch.linalg.norm(x, p, dim=1).detach()
```
added assert and docstring

AlekseySh commented on oml/utils/misc_torch.py:
```diff
@@ -103,6 +114,14 @@ def _check_is_sequence(val: Any) -> bool:
+
 class OnlineCalc(ABC):
     """
     The base class to calculate some statistics online (on the steam of values).
```
fixed
